### PR TITLE
fix: semver like titles

### PR DESCRIPTION
**Pull request by Eomm**
fixes https://github.com/fastify/github-action-merge-dependabot/issues/124

It should support submodules as well (the #98 PR adds the `any` option to skip this check)

### CODE REVIEW

**Review by simoneb — COMMENTED** [review]
Do we have a test which checks that even when specifying a target, submodules are handled? I'm not actually sure what would be the expected behavior though, maybe it's worth deciding and documenting?

**Comment by Eomm** [comment]
> Do we have a test which checks that even when specifying a target, submodules are handled? I'm not actually sure what would be the expected behavior though, maybe it's worth deciding and documenting?

Added a new test to check it.
I think the behaviour is consistent within the documentation now

File: test/action.test.js [hunk 1]
```diff
@@ -263,3 +263,29 @@ tap.test('should call external api for github-action-merge-dependabot major rele
   t.ok(stubs.logStub.logWarning.calledOnce)
   t.ok(stubs.fetchStub.calledOnce)
 })
+
+tap.test('should check submodules semver when target is set', async t => {
```
Comment (simoneb) [inline test/action.test.js:267]: this is not what submodules look like. submodules have a commit sha in the pr title (iirc) https://github.com/fastify/github-action-merge-dependabot/issues/95

Comment (simoneb) [reply test/action.test.js:267]: @Eomm lmk if the comment is clear. It's about the title of the PR you have in this test not matching what happens when a version of a submodule tries to be bumped. You only get a SHA in that case

Comment (Eomm) [reply test/action.test.js:267]: Yeah, I'm on it. Let me push the wip code

I'm trying to ignore the case:

- the PR title is not a semver and the user set the `target` value

this check is a bit tricky due the `prerelease` stuff and the coercion.

Comment (simoneb) [reply test/action.test.js:267]: we don't have to support _everything_, it would be a good step if we didn't break anything of the existing and we document the behavior for edge cases so we know what we can work on next

**Review by mcollina — APPROVED** [review]
lgtm

File: test/action.test.js (hunk 1 above)
Comment (Eomm) [reply test/action.test.js:267]: I have implemented the most simple check, using this PR as reference:

https://github.com/sitiom/dotfiles/pull/1

File: README.md [hunk 2]
```diff
@@ -43,6 +43,9 @@ Possible options are:
 
 For more details on how semantic version difference is calculated please see [semver](https://www.npmjs.com/package/semver) package.
 
+If you set a value other than `any`, PRs that are not semantic version compliant are skipped.
+An example of a non-semantic version is a commit hash.
```
Comment (simoneb) [inline README.md:47]: looks good, can we add just one piece to this sentence to clarify when this may happen? e.g. : when using git submodules

File: src/checkTargetMatchToPR.js [hunk 3]
```diff
@@ -11,11 +12,23 @@ const checkTargetMatchToPR = (prTitle, target) => {
   if (!match) {
     return true
   }
-  const diff = semverDiff(match[1], match[2])
+
+  const [, from, to] = match
+
+  if(hasBadChars(from) || hasBadChars(to)) {
+    return false
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:19]: let's log something when this happens. also not clear to me what this means to be honest

File: src/checkTargetMatchToPR.js [hunk 4]
```diff
@@ -11,11 +12,23 @@ const checkTargetMatchToPR = (prTitle, target) => {
...
+  const diff = semverDiff(semverCoerce(from), semverCoerce(to))
 
   return !(
     diff &&
     semanticVersionOrder.indexOf(diff) > semanticVersionOrder.indexOf(target)
   )
 }
+
+function hasBadChars(version) {
+  return /`/.test(version)
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:31]: continuing from the earlier comment, it's not really clear what this means and when it may happen

Comment (simoneb) [reply src/checkTargetMatchToPR.js:31]: ah ok I see it now looking at the test below. I'm not sure this is a very robust way to check if we're dealing with SHAs though. What if gh decides to change dependabot and also put backticks around semver version numbers?

Comment (Eomm) [reply src/checkTargetMatchToPR.js:31]: Yes, it is true, we depend on the title format even more than before.
I agree with that.

the issue is that the semver coerce returns a semver object for the hash string too.
So we must understand if the string we are going to coerce could be a reasonable semver (like `2` or `2.3`)
We can't check for alphanumeric value, because of the pre-release format.

I was working on a more robust solution but some cases were not covered. Should I restore that code and go for this edge case?

Comment (simoneb) [reply src/checkTargetMatchToPR.js:31]: I'm happy with this change but let's be more explicit about what this does because nothing in the code captures that this is to handle SHAs. If needed, let's add a comment. Then this PR is good to go

**Review by simoneb — APPROVED** [review]
